### PR TITLE
Support willReadFrequently for CanvasRenderingContext2D

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### fea3b372b1651a492a3a7f314b3120b5da15518b
<pre>
Support willReadFrequently for CanvasRenderingContext2D
<a href="https://bugs.webkit.org/show_bug.cgi?id=244117">https://bugs.webkit.org/show_bug.cgi?id=244117</a>

Reviewed by Darin Adler.

It&apos;s used to indicate there will be frequent get image data operations,
in which case an accelerated canvas can be slower than unaccelerated
one. This patch turns off hardware acceleration for the 2D canvas when
willReadFrequently is true in settings.
See <a href="https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently">https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently</a>

Test: fast/canvas/canvas-will-read-frequently.html

* LayoutTests/TestExpectations: Skip the test in all platforms.
* LayoutTests/fast/canvas/canvas-will-read-frequently-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-will-read-frequently.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/the-offscreen-canvas/2d-getcontext-options.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/the-offscreen-canvas/2d-getcontext-options.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-canvas-element/2d-getcontext-options-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-full-expected.txt:
* LayoutTests/inspector/canvas/recording-2d-memoryLimit-expected.txt:
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-full-expected.txt:
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-memoryLimit-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/inspector/canvas/context-attributes-expected.txt:
* PerformanceTests/Canvas/getImageDataCPU.html: Added.
* PerformanceTests/Canvas/getImageDataGPU.html: Added.
* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::allocateImageBuffer const):
* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::willReadFrequently const):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::willReadFrequently const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.idl:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::buildObjectForCanvasContextAttributes):

Canonical link: <a href="https://commits.webkit.org/277494@main">https://commits.webkit.org/277494@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1a500a4a075126421f3ff605ba7100a5dcfb5dce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50538 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50446 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43818 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50070 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24437 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38873 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20170 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22095 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42443 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5812 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41049 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44119 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52340 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47256 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19140 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46179 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24072 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45217 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10539 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24861 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54754 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23793 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11243 "Passed tests") |

### CODE REVIEW

**Comment by smfr** [comment]
Is there evidence that this is a performance win in any perf test that we can construct?

**Comment by carlosgcampos** [comment]
It depends on the hardware, but while working on Skia I noticed that, for example in MotionMark, the images test is always faster on the CPU, you can try by just running the test with 2D canvas acceleration setting disabled. Then I found that `willReadFrequently` setting while reading chromium code, and I thought it would be good for WebKit to also handle it.

**Comment by carlosgcampos** [comment]
I've just realized that images test in MotionMark is not a good one, it worked in Skia before because we always accelerated the canvas, but now we use the CPU for small canvases and that test crates canvases of 50x50 that are never accelerated in any case.

File: Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.h [hunk 1]
```diff
@@ -33,6 +33,7 @@ namespace WebCore {
 
 struct CanvasRenderingContext2DSettings {
     bool desynchronized { false };
+    bool willReadFrequently { false };
```
Comment (darinadler) [inline Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.h:36]: Why not keep this in the same order as the IDL?

**Review by darinadler — APPROVED** [review]
I wish the change log was clearer on this point: "will read frequently" **turns off hardware acceleration**. We say "accelerated canvas **can** be slower", which is true, but doesn’t go far enough: we should say what browser engines are expected to do, not just document the rationale.

**Comment by carlosgcampos** [comment]
Thanks for the review @darinadler, I'll wait until @smfr confirms it's ok before landing. I can also make it Skia specific for now, if you are not sure you want this in Apple ports.

**Comment by annevk** [comment]
@carlosgcampos @smfr replied on Slack yesterday asking again for evidence this improves actual workloads. Not enabling it for Apple ports would probably also work.

**Comment by carlosgcampos** [comment]
The updated patch includes two perf tests based on the ones I found in chromium. 

* With cairo (only CPU is supported, so willReadFrequently is just ignored)
```
getImageDataCPU.html
:Runs -> [959.9999999999997, 966.1835748792271, 962.6955475330926, 968.5230024213081, 965.0180940892631] runs/s

getImageDataGPU.html
:Runs -> [963.855421686747, 982.8009828009828, 975.609756097561, 970.8737864077669, 973.2360097323601] runs/s
```

* With Skia
```
getImageDataCPU.html
:Runs -> [969.6969696969699, 939.9477806788515, 972.0534629404617, 975.609756097561, 976.8009768009773] runs/s

getImageDataGPU.html
:Runs -> [312.5, 313.3159268929504, 314.5478374836173, 320.5918618988903, 310.4786545924969] runs/s
```

With cairo we get the same results, as expected, but with Skia we get much better results with the CPU.

**Comment by annevk** [comment]
@carlosgcampos would you be able to get macOS results as well? If not I might be able to take a look at some point.

**Comment by carlosgcampos** [comment]
I don't have a mac, I can get access to one remotely if needed, but it will take some time.

**Comment by carlosgcampos** [comment]
@nikolaszimmermann ran the tests in his mac with these results:

```
RESULT Canvas: getImageDataCPU: Runs= 1017.0378024739535 runs/s
median= 1009.4641242182225 runs/s, stdev= 18.7276560741 runs/s, min= 990.0990099009902 runs/s, max= 1049.8687664041995 runs/s

RESULT Canvas: getImageDataGPU: Runs= 441.53215587584225 runs/s
median= 408.993946758919 runs/s, stdev= 80.3540738042 runs/s, min= 381.40643623361143 runs/s, max= 639.1478029294275 runs/s
```

So, much better results in Mac too with the CPU.

**Comment by kkinnunen-apple** [comment]
Thanks!
The drawing is memset(p, 0, 1024*1024*4); ... CPUS are awfully fast filling memory that way..
It's not really toggle "I know I will read frequently", rather "I know my drawing is trivial"...
It's good data point, but it'd be more compelling to understand a real-world case for it.
If you want to proceed with the change now, I'd suggest you ifdef COCOA out of it and we'd try to find real-world use-cases.

File: Source/WebCore/html/CanvasBase.cpp [hunk 2]
```diff
@@ -334,18 +334,16 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer() const
...
         bufferOptions.add(ImageBufferOptions::Accelerated);
-
-    auto colorSpace = DestinationColorSpace::SRGB();
-    auto pixelFormat = PixelFormat::BGRA8;
-
-    if (auto* context = renderingContext()) {
+    if (context)
         bufferOptions = context->adjustImageBufferOptionsForTesting(bufferOptions);
-        colorSpace = context->colorSpace();
-        pixelFormat = context->pixelFormat();
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/CanvasBase.cpp:347]: E.g. this hunk could be preserved and
```
#if !PLATFORM(COCOA)
        if (context->willReadFrequently())
              bufferOptions.remove(..Accelerated..);
#endif
```

Comment (carlosgcampos) [reply Source/WebCore/html/CanvasBase.cpp:347]: Ok, but I took a different approach that I find easier to follow.

**Pull request merged**